### PR TITLE
feat(cablecar): change the default mimetype

### PR DESCRIPTION
**Pull request by existentialmutt**
## Status

* Needs Review

## Additional Notes

This PR makes some adjustments to the CableCar plugin to have it use response Content-Type as the determining factor of whether to run CableReady operations. The MIME type defaults to the CR-specific `application/vnd.cable-ready.json` but can be changed by the configuration parameter we introduced yesterday.

This allows the client to request CableCar data from the server programatically using mrujs.fetch without registering a separate listener.

~~I did find that when using a custom content type, fetchResponse.isJson and fetchResponse.responseJson weren't working correctly, so I've switched to using fetchResponse.responseText.  I thought you fixed that last week but maybe there's another issue?~~  Thanks for fixing in main!

Note that the server must be able to respond with the correct content-type.  @leastbad and I were discussing this yesterday and it sounds like a dedicated cable_ready content type is a good idea.  It's a one-line change to have CableReady do that and I'm about to submit the PR for that as well.  I'll post the url in the comments below.

Finally, if you want to see how this all fits together, I developed it against my cable_modal proof of concept.  [There's a demo branch up here](https://github.com/existentialmutt/cable_modal_poc/tree/mrujs_cablecar_plugin_pr_tests).  Making these changes allows that code to focus on it's own domain and offload all operation fetching and processing to the mrujs cablecar plugin.

### CODE REVIEW

**Comment by existentialmutt** [comment]
CableReady PR for registering the MIME type is here https://github.com/stimulusreflex/cable_ready/pull/140

**Comment by existentialmutt** [comment]
here's an example of how @leastbad 's CableCar example gist would change with this modification

https://gist.github.com/existentialmutt/296f9745b17383c4549d286dc2d386dd

**Comment by ParamagicDev** [comment]
@existentialmutt the json issue shouldve been fixed here:

https://github.com/ParamagicDev/mrujs/blob/668991ac3b47047d3c8af18d437c823cb7174ea6/src/http/fetchResponse.ts#L79

**Comment by existentialmutt** [comment]
> @existentialmutt the json issue shouldve been fixed here:
> 
> https://github.com/ParamagicDev/mrujs/blob/668991ac3b47047d3c8af18d437c823cb7174ea6/src/http/fetchResponse.ts#L79

yeah that should do it.  Thanks for the pointer I'll dig around and see if I can find out why it doesn't like that.

**Comment by existentialmutt** [comment]
@ParamagicDev fix in main worked great.  Back to using responseJson

**Comment by leastbad** [comment]
Hey, forgive the drive-by code review...

```rb
respond_to do |format|
  format.cable_ready do
    render operations: cable_car.console_log(message: "Destroy #{params[:id]}!")
  end
end
```
<
```rb
render operations: cable_car.console_log(message: "Destroy #{params[:id]}!")
```
By which I mean... I'm all for options, but please don't make the `respond_to` block structure mandatory. So much ceremony!

**Comment by existentialmutt** [comment]
@leastbad That's a fair point.  I think this approach may require some kind of format indication but I'm going to look into what our options are.

**Comment by existentialmutt** [comment]
@leastbad we could make respond_to blocks optional by setting the Content Type as part of `render operations:`

https://github.com/stimulusreflex/cable_ready/pull/140/commits/c981d8b8d521d13849bc1035f5dcca313278a215

**Comment by leastbad** [comment]
Cooool! I can't be the one to approve it (because again, not my wheelhouse) but LGTM. You should convert that to a PR.

File: src/plugins/cableCar.ts [hunk 1]
```diff
@@ -43,6 +43,7 @@ export class CableCar {
     document.addEventListener('DOMContentLoaded', this.boundScanner)
     document.addEventListener('turbolinks:load', this.boundScanner)
     document.addEventListener('turbo:load', this.boundScanner)
+    document.addEventListener(AJAX_EVENTS.ajaxComplete, this.boundPerform)
```
Comment (ParamagicDev) [inline src/plugins/cableCar.ts:46]: Good call. No reason to have these element level when they all bubble up to the document

File: src/plugins/cableCar.ts [hunk 2]
```diff
@@ -29,7 +29,7 @@ export class CableCar {
     this.observer = new MutationObserver(this.boundScanner)
     this.elements = []
     this.cableReady = cableReady
-    this.mimeType = (mimeType ?? 'json')
+    this.mimeType = (mimeType ?? 'application/vnd.cable-ready.json')
```
Comment (ParamagicDev) [inline src/plugins/cableCar.ts:32]: Will this be the standard mimetype for cable ready?

File: .gitignore [hunk 3]
```diff
@@ -145,3 +145,6 @@ test/rails/dummy/yarn-error.log
 test/rails/dummy/yarn-debug.log*
 test/rails/dummy/.yarn-integrity
 **/*.DS_Store
+
+# nodenv version file
+.node-version
```
Comment (ParamagicDev) [inline .gitignore:150]: YOU THINK YOU CAN COME IN HERE AND MESS WITH MY CAREFULLY CURATED `.gitignore` ?!!

File: src/plugins/cableCar.ts [hunk 4]
```diff
@@ -53,13 +54,13 @@ export class CableCar {
...
   disconnect (): void {
     this.elements.forEach((element: ExtendedElement) => {
-      element.removeEventListener(AJAX_EVENTS.ajaxComplete, this.boundPerform)
       element.observer?.disconnect()
     })
 
     document.removeEventListener('DOMContentLoaded', this.boundScanner)
     document.removeEventListener('turbolinks:load', this.boundScanner)
     document.removeEventListener('turbo:load', this.boundScanner)
+    document.removeEventListener(AJAX_EVENTS.ajaxComplete, this.boundPerform)
```
Comment (ParamagicDev) [inline src/plugins/cableCar.ts:63]: 👍

**Review by ParamagicDev — APPROVED** [review]
Just one question on mime-types, otherwise this is good to go!

I worry that by swapping the default mime-type it may confuse users like @leastbad who just treat it as normal json.

**Comment by leastbad** [comment]
I think that my fundamental hope/wish was that `render operations: cable_car` would be `application/json` unless someone *also* uses a `respond_to` block, which would set it to `cable_ready`. I don't know if that works! This area isn't my forte.

**Comment by existentialmutt** [comment]
Fair enough @leastbad!  Yes, I'm suggesting we change the default Content Type for cablecar to application/vnd.cable-ready.json and I realize that's not a small ask.

I'm advocating this approach because Content Type is one of the few CORS/CSP approved signals a request can use to indicate a specific purpose.  This allows the mrujs cablecar plugin, cable_modal, and other general purpose utilities built around cable_car to register global event handlers on document that won't step on other json requests that could be on a complex page.

I realize this adds something to think about for anyone who wants to use cablecar with json content type and be done.  My hope is that by making application/vnd.cable-ready.json the default it should be pretty unobtrusive. Everything works as it did before, there's just a different Content Type on the response now.  If your client-side code doesn't look at response Content Type nothing needs to change (btw mrujs is smart enough to recognize application/vnd.cable-ready.json as a JSON content type so it's parsing helpers are still available)

It's also entirely possible I have an XY problem going on here and this isn't the best solution. The specific thing I'm trying to achieve is to allow the mrujs cablecar plugin to register a CableReady.perform event handler that clearly differentiates cablecar requests from other json without having those requests be tied to a specific DOM element.  I'm game for other solutions.

Finally, I'm very much on board with the idea of "CableReady, the Format" as something that can be easily ported to other languages and frameworks. I think establishing a dedicated MIME type for CableReady signals reliability and maturity on that front.  I'm also hearing there's a desire for application/json and I want to better understand where that comes from so we can make a solution that covers as many bases as possible.

**Comment by leastbad** [comment]
Really appreciate the explainer. I'm still parsing.

I am fine with switching MIME types on the client. It seems that it would be effectively transparent to users of, say, an updated mrujs, right?

Lower level, if I'm doing a fetch call do I have to specify an Accepts header in this scheme?

On the server, I understand that we can just set the new default MIME type in the renderer for `operations`, which I already saw and again, I'm fine with - it seems transparent.

What I'm trying to avoid is a scenario where someone has to use the `respond_to` block structure in order for all of this to work - increasing the line count and cognitive complexity of sending CableReady JSON from one line to 4-5.

Hopefully, I'm just asking [another] naive question and fretting over nothing...?

**Comment by ParamagicDev** [comment]
So theoretically we could do what Turbo does. Turbo has their Turbo Stream mime type. So what they do under the head is add prepend it to the accept header.

If you image the Accept header usually looks like this:

`{Accept: "application/html"}`

With this line of code here:

https://github.com/hotwired/turbo/blob/a47e5b62646fcf4e285dbc7308e91c752aec7369/src/core/drive/form_submission.ts#L123

It goes from the above to this:

```js
{Accept: "turbo/stream.mime, application/html"}
```

This means that Turbo stream responses take precendence over HTML. We could do something similar here.

Mrujs' default mime type is: "*/*"

We could make it: "text/cablecar.json, */*" which means if the server has both CableReady and HTML views for a route, itll prioritize the cableready response with a respond_to block.

**Comment by leastbad** [comment]
I might actually be more confused, now... sorry! I thought that maybe if I read this again, later, I'd get it. 🤦 

My goal is keeping the API short and easy to use on both the client and the server. Right now, I can use `fetch` to request a path, and in my controller I have a 1-liner that can return a CableReady compliant JSON string that on the client, I can interpret as JSON and pass directly to CableReady. It's simple, requires no extra code, and it works great, today.

I am frustrated that we're discussing taking something that is both simple and **already working**, adding more complexity to it... to achieve the *same fundamental outcome*.

Rafe said:
> I realize this adds something to think about for anyone who wants to use cablecar with json content type and be done. My hope is that by making application/vnd.cable-ready.json the default it should be pretty unobtrusive.

This is the simplest `fetch` GET call that works today:

```js
fetch('/foo')
  .then(r => r.json())
  .then(CableReady.perform)
```

Does that have to change? If the answer is yes, then we should go back to the drawing board.

Adding **optional** things to the above will always be 100% fine, but fundamentally fetching JSON and feeding it to CableReady should always be this simple. In fact, this conversation makes me want to go into my Webpacker plugins and create a global function called `json`...

```js
function json(data) { return data.json() }
```

... just so that I can...

```js
fetch('/foo')
  .then(json)
  .then(CableReady.perform)
```

Moving on to the server side... thanks to the [custom renderer](https://github.com/stimulusreflex/cable_ready/blob/3b0c918ae8f848a9281025b4a7ee5838a128edaa/lib/cable_ready.rb#L32), we can reply to our fetch request with a 1-line controller action:

```rb
render operations: cable_car.blah
```

I'm now going to cut-and-paste myself without shame:

> Does that have to change? If the answer is yes, then we should go back to the drawing board. Adding **optional** things to the above will always be 100% fine, but fundamentally fetching JSON and feeding it to CableReady should always be this simple.

What do I mean by optional things? For starters, I actually don't think I care about the MIME type so long as it doesn't change the `fetch` call. If you want to change the MIME type because it advances the cause, I'll let you folks sort that out so long as you don't have to add headers to every `fetch` call until the end of time.

Another optional thing is this `cable_ready` format:

```rb
respond_to do |format|
  format.cable_ready do
    render operations: cable_car.blah
  end
end
```
Again, this looks like it could be really useful! Just... don't make this the only way to return CableReady JSON. So long as the `render operations: cable_car.blah` syntax still works without having to think about MIME types, I'm excited that you folks are pushing things forwards.

Heck, you could even introduce a second renderer helper, if that solves any problems:

```rb
respond_to do |format|
  format.cable_ready do
    render cable_ready: cable_car.blah
  end
end
```

That seems like it could make a lot of good sense, semantically. If you render `operations`, you get a string that can be parsed as JSON that shows up with standard `application/json`, but if you render `cable_ready`, you get a magical string that can be parsed as JSON that shows up as `application/vnd.cable-ready.json` or whatever it was that Rafe proposed earlier. (LGTM!)

Now, it's possible that I'm just cranky and regurgitating good ideas y'all have already said and we are in violent agreement and I'm embarrassing myself. In fact, I hope that some version of that is true. I'm fine with it, if it means that we ultimately get to keep things simple for the users. 💙

**Comment by ParamagicDev** [comment]
@leastbad So let me pull this back.

Lets say I have `/foo` I have `app/views/foo/index.html.erb` but I want the cableready JSON response for that instead.

Today, you would have to do this to make this happen:

```js
mrujs.fetch('/foo', { headers: "application/json" })
  .then(response => response.json())
  .then(CableReady.perform)
```

Really at the end of this day this is to fix if I have 3 conflicting routes:

```js
foo.json
foo.html
foo.cable_ready
```

I need some way to seperate a jbuilder json response from a cableready json response.

I could even have views that look like this:

```js
foo.json.jbuilder
foo.html.erb
foo.cable_ready.erb
```

And based on the `Accept` header will auto send back the one I want.

God this is so confused to explain...

**Comment by leastbad** [comment]
Okay, but it's important to differentiate that `mrujs.fetch != fetch`.

Right now, I can `fetch` exactly as I described, without needing to set an `Accept` header.

**Comment by ParamagicDev** [comment]
Let me back up, assuming you have a the following:

if you have two views (a JSON and an HTML view) on the route, the HTML will always be sent back without an `Accept` header to make the JSON take precedence over HTML.

**Comment by leastbad** [comment]
Okay, I think I'm all clear, here, finally.

I left a note on https://github.com/stimulusreflex/cable_ready/pull/140 that I have since retracted.

I now understand that you only ever need to use a `respond_to` block if you have more than one MIME type available at a given route.

**Comment by existentialmutt** [comment]
Appreciate the spirited discussion guys!

@leastbad thank you for clarifying where you stand.  The "ain't broke, don't fix it" approach makes perfect sense and I appreciate you hanging in here for a feature whose benefits aren't exactly obvious.

I want to be sure of the behavior too and I've been running some tests tonight with various combinations of Accept Headers set on the request and Content Types set on the response.  The good news is almost every combination works and will render CableReady operations!

Screenshot 2021-07-05 at 01-26-03 CablecarMimeTest

What you're looking at is running through the matrix of request/Accept Headers X methods of setting response content_type:

The different Accept headers are:
- Default ("*/*")
- application/json
- application/vnd.cable_ready.json
- application/vnd.cable_ready.json, application/json
- Extra-simple GET fetch (@leastbad I threw in the exact method you described above just to be sure)

The methods of setting response content type are:
- Default Content Type: nothing special, just render: operations
- Set Response Content Type: call `response.content_type = Mime[:cable_ready]` in the controller before rendering
- Use Respond To Block: use a `respond_to {|format| format.cable_ready {...` block in the controller

As you can see, the only combination that refuses to execute cableready operations is when you specifically request json but tell the controller to only render cable_ready with a respond_to block.  That one failing to perform actually makes sense to me, and it's encouraging that everything else allows the operations to go through!

Also encouraging is that I tried switching back and forth between mrujs fetch and vanilla fetch and they have identical results.

Code is here if you want to review: https://github.com/existentialmutt/cablecar_mime_test

In general what I'm taking away from this is that browsers and Rails are both pretty forgiving around matching MIME types.

**Comment by leastbad** [comment]
You rock, Rafe! This is high-quality rigor, not that you needed to endear yourself further.

My major conceptual stumbling block is something Konnor managed to finally break up after over a decade of blissful ignorance: none of this particularly matters unless you're serving different response types. Since I almost always use standalone routes for Ajax calls, I've just never had a proper 💡 moment with this mechanism.

Now, I see... and I suspect it would be hard to un-see. 👁️

**Comment by existentialmutt** [comment]
Haha, awesome! This has been educational for me as well

**Review by ParamagicDev — APPROVED** [review]

**Pull request merged**